### PR TITLE
JIT: Use `jitstd::utility::scoped_code` to unmark trees during lowering

### LINKED ISSUE #105971: JIT: checked-release asmdiffs for libraries_tests.run.linux.arm64.Release.mch
4 diffs overall, perhaps all the same or in a similar method

See https://dev.azure.com/dnceng-public/public/_build/results?buildId=765225&view=results

```
[11:10:01] ISSUE: <ASM_DIFF> main method 617927 of size 380 differs
[11:03:31] ISSUE: <ASM_DIFF> main method 710437 of size 380 differs
[11:03:31] ISSUE: <ASM_DIFF> main method 710603 of size 380 differs
[11:03:31] ISSUE: <ASM_DIFF> main method 710588 of size 380 differs
```

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Fixes #105971. `Lowering::TryMakeIndirsAdjacent` and `Lowering::TryMakeIndirAndStoreAdjacent` require that no `GenTree` nodes are marked upon entry, and that marked nodes are cleared upon exit, but in the case of the first method, not every exit path was unmarking the tree. `jitstd::utility::scoped_code` makes this invariant clear.

### CODE REVIEW

**Comment by amanasifkhalid** [comment]
cc @dotnet/jit-contrib, @jakobbotsch PTAL. Thanks!

**Review by jakobbotsch — APPROVED** [review]
Thanks for fixing this!

File: src/coreclr/jit/lower.cpp [hunk 1]
```diff
@@ -9355,6 +9358,12 @@ bool Lowering::TryMakeIndirsAdjacent(GenTreeIndir* prevIndir, GenTreeIndir* indi
 
 #endif
 
+    // Unmark tree when we exit the current scope
+    auto code = [this, indir] {
+        UnmarkTree(indir);
+    };
+    jitstd::utility::scoped_code<decltype(code)> finally(code);
```
Comment (teo-tsirpanis) [inline src/coreclr/jit/lower.cpp:9365]:
```suggestion
    jitstd::utility::scoped_code finally([this, indir] {
        UnmarkTree(indir);
    });
```

File: src/coreclr/jit/lowerarmarch.cpp [hunk 2]
```diff
@@ -1197,6 +1200,12 @@ bool Lowering::TryMakeIndirAndStoreAdjacent(GenTreeIndir* prevIndir, GenTreeLclV
 
 #endif
 
+    // Unmark tree when we exit the current scope
+    auto code = [this, store] {
+        UnmarkTree(store);
+    };
+    jitstd::utility::scoped_code<decltype(code)> finally(code);
```
Comment (teo-tsirpanis) [inline src/coreclr/jit/lowerarmarch.cpp:1207]:
```suggestion
    jitstd::utility::scoped_code finally([this, store] {
        UnmarkTree(store);
    });
```

**Review by teo-tsirpanis — COMMENTED** [review]
There's a chance (I'm not 100% sure, I haven't tried it) that C++ can deduce the template argument, which can enable inlining the `auto code = ` variable.

**Comment by amanasifkhalid** [comment]
> There's a chance (I'm not 100% sure, I haven't tried it) that C++ can deduce the template argument, which can enable inlining the auto code =  variable.

Thank you for the suggestion! I tried this, and it didn't compile... I don't think this type deduction works when it's used on a templated class, at least in C++11.

**Comment by teo-tsirpanis** [comment]
Oh you are on C++11, this feature was [added in C++17](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction). 😔

**Comment by amanasifkhalid** [comment]
> Oh you are on C++11, this feature was [added in C++17](https://en.cppreference.com/w/cpp/language/class_template_argument_deduction). 😔

Yeah, I'm not sure if we have any plans to upgrade in the near future, though it's helpful to know this is another pattern we could clean up if we did upgrade.

**Pull request merged**